File: csharp/2015/day04b/Program.cs
using System;
using System.Security.Cryptography;
using System.Text;

class Program
{
    static void Usage()
    {
        Console.Error.WriteLine("usage: dotnet run <key>");
        Environment.Exit(1);
    }

    static int Process(string key)
    {
        using MD5 md5 = MD5.Create();
        int n = 1;

        while ( true )
        {
            string tryKey = key + n.ToString();
            byte[] inputBytes = Encoding.UTF8.GetBytes(tryKey);
            byte[] hashBytes = md5.ComputeHash(inputBytes);

            StringBuilder sb = new StringBuilder();
            foreach ( byte b in hashBytes )
            {
                sb.Append(b.ToString("x2"));
            }
            string digest = sb.ToString();

            if ( digest.StartsWith("000000") )
            {
                break;
            }
            n++;
        }

        return n;
    }

    static void Main(string[] args)
    {
        if ( args.Length < 1 )
        {
            Usage();
        }

        string key = args[0];
        int result = Process(key);
        Console.WriteLine("result = " + result);
    }
}

File: fsharp/2015/day04b/Program.fs
open System
open System.Security.Cryptography
open System.Text

let usage() =
    Console.Error.WriteLine("usage: dotnet run <input file>")
    Environment.Exit(1)

let processKey(key: string) =
    use md5 = MD5.Create()
    let rec f n =
        let tryKey = key + n.ToString()
        let inputBytes = Encoding.UTF8.GetBytes(tryKey)
        let hashBytes = md5.ComputeHash(inputBytes)
        let hexDigest = hashBytes |> Array.map (fun b -> b.ToString("x2")) |> String.Concat
        if hexDigest.StartsWith("000000") then
            n
        else
            f (n + 1)
    f 1

[<EntryPoint>]
let main args =
    if args.Length < 1 then
        usage()

    let key = args.[0]
    let result = processKey key
    Console.WriteLine("result = " + result.ToString());
    0 
